crates/rollup/src/driver.rs

```diff
@@ -147,14 +168,164 @@ where
         )
     }
 
-    /// Starts the Hera Execution Extension loop.
+    /// Advance the pipeline to the next L2 block.
+    async fn step(&mut self, pipeline: &mut RollupPipeline<CP, BP, L2CP>) -> Result<()> {
+        let (l2_tip, l1_origin) = self.cursor.tip();
+        let _ = pipeline.step(l2_tip).await;
+        self.cursor.advance(l1_origin, l2_tip);
+
+        unimplemented!()
+    }
+
+    async fn handle_notification(
+        &mut self,
+        notif: ExExNotification,
+        pipeline: &mut RollupPipeline<CP, BP, L2CP>,
+    ) -> Result<()> {
+        if let Some(reverted_chain) = notif.reverted_chain() {
+            // The reverted chain contains the list of blocks that were invalidated by the
+            // reorg. we need to reset the cursor to the last canonical block, which corresponds
+            // to the block before the reorg happened.
+            let last_canonical_block = reverted_chain.first().number - 1;
+
+            // Find the last known L2 block that is still valid after the reorg,
+            // and reset the cursor and pipeline to it.
+            let (l2_safe_tip, l2_safe_tip_l1_origin) = self.cursor.reset(last_canonical_block);
+
+            warn!("Reverting derivation pipeline to L2 block: {}", l2_safe_tip.number);
+            if let Err(e) = pipeline.reset(l2_safe_tip, l2_safe_tip_l1_origin).await {
+                bail!("Failed to reset pipeline: {:?}", e);
+            }
+        }
+
+        if let Some(committed_chain) = notif.committed_chain() {
+            let tip_number = committed_chain.tip().number;
+            // self.chain_provider.commit_chain(committed_chain);
+
+            if let Err(err) = self.ctx.send_event(ExExEvent::FinishedHeight(tip_number)) {
+                bail!("Critical: Failed to send ExEx event: {:?}", err);
+            }
+        }
```

Comment:
I think we ought to match on the `ExExNotification`. It's not marked as `non_exhaustive`, which leads me to believe new variants won't be added without a major (maybe minor) version bump.

Additionally, it looks like we're missing the case where the chain is reorged. See [`ExExNotification::ChainReorged`](https://github.com/paradigmxyz/reth/blob/51524f2534bf9b28c87be4321b6dd41378caf2f8/crates/exex/types/src/notification.rs#L14-L20). In the re-orged case, it looks like we'll want to add the new chain to the local provider, and then then _also_ walk back the old chain like logic you wrote above does for the `ChainReverted` notification.